rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+main operation batch-reduce GEneral Matrix Multiply operation (aka
+brgemm), is a very flexible operation that can be used in a variety of
+DNN operators. It is defined as folllow:
+
+```math
+D = \beta C + \alpha \sum_i A_i \cdot B_i + bias
+```
+
+with 
+- $A_i$ a set of matrices of dimension $M \times K$
+- $B_i$ a set of matrices of dimension $K \times N$
+- D and C matrices of dimension $M \times N$
+- bias a vector of dimension $N$.
+
+This proposal discusses exposing these sequential, basic building
+blocks externally for end-users to conveniently be able to write their
+own optimized custom primitives/routines. These can be used in their
+own applications and/or contributed back behind the oneDNN graph API.
+
+As a starter, we will expose only simple transforms and brgemm
+routines.
+
+## Memory descriptors, memory objects and/or pointers ?
+
+In general, we need to express memory shapes and memory objects, and
+for this block level API, we will have are three considerations:
+- allowing arbitrary strides between blocks ($A_i$, $B_i$ in
+  particular) for maximum flexibility,
+- Support for metadata (e.g. zero-point handling) and
+  hardware-specific layouts,
+- runtime overheads.
+
+### Arbitrary strides between A and B blocks
+
+To make the API flexible, we want to allow arbitrary strides between A
+and B blocks as showed in the picture:
+![](brgemm_pic.png)
+
+This is necessary on multiple occasions, a few examples being:
+- for convolutions, where blocks can overlap depending on covolution strides
+- for LSTM/GRU where weights for each gate can live in separate memory addresses.
+- for models with embedding lookup, where vectors can be passed as separate blocks.
+
+Using arbitrary strides between blocks is unfortunately not possible
+if we encapsulate all $A_i$ blocks (resp $B_i$ blocks) of a brgemm
+operation in a dnnl::memory object, since this will force a fixed
+stride between each block.
+
+Another option would be to take an array of dnnl::memory objects, each
+containing a single block of A/B. This would require the user to call
+`dnnl::memory::set_data_handle()` on each of these objects before
+brgemm call. It would incur some unnecessary overhead, but should
+not be anything substantial.
+
+However, one might wonder why use memory descriptors at all and not
+just take pointers directly as input to brgemm::execute
+call. Internally we support multiple flavors:
+- array of pointers: here the user will have to do pointer arithmetic
+  before passing those pointers to brgemm::execute call.
+- base pointer + arrray of offsets: here the pointer arithmetic will
+  be handled internally behind brgemm::execute call.
+- base pointer + fixed stride: this flavor has limited application and
+  is typically used for matmul implementation. Given its limited
+  applicability, it seems unnecessary to expose it.
+  
+To make the public brgemm API simpler, we propose to expose only the array
+of pointers flavor.
```

Comment:
+ @ankalinin @nivas-x86 to comment on that.

Could you clarify what you mean by compile-time constants? Do you mean brgemm object creation time?

In general, a few things:
- if brgemm object creation takes pointers as parameters, it will severely reduce the possibility to reuse brgemm objects/kernels across an application, and introduce jitting overheads. With that respect, base pointer + array of offsets might be more reusable.
- even if the pointers are constant, we typically still need to put those pointers in register. IIRC, the fastest way to put a 64 bit value in register on x64 is to actually do a mov from memory. In which case whether the pointer is constant or not should not matter. Do I miss something here?
- the pointers are typically not constant at brgemm object creation-time. However, the offsets might be. In that case, the base pointer + array of offset would be the fastest option, as it can remove some overhead related to pointer arithmetic (by reusing buffer of offsets without resetting it before every brgemm call). 

Actually this discussion is making me want to revise to using base pointer + array of offsets :).
Adding @ankalinin and @nivas-x86 for discussion.

Reply:
> the pointers are typically not constant at brgemm object creation-time. However, the offsets might be. In that case, the base pointer + array of offset would be the fastest option, as it can remove some overhead related to pointer arithmetic

Yup, that was my line of thinking. Take the inference case as an example: the compiler will be able to prepopulate the weight matrix offsets in advance, cutting the cost of address computation. While good on paper, it is unclear to me whether that would make any real difference, hence the question.

Reply:
Fair observation. The usages I saw in IPEX, openVINO and oneDNN all rely on the pointer based flavor.
I will let @nivas-x86 and @ankalinin comment about potential speedups when using precomputed offsets.

Reply:
Graph compiler uses `base pointer + fixed stride` flavor for matmul implementation and `array of pointers` for convolution. For most DL compilers, it's not easy to describe pointer arithmetic. Take brgemm in Linalg dialect(https://mlir.llvm.org/docs/Dialects/Linalg/#linalgbatch_reduce_matmul-linalgbatchreducematmulop) as example, `base pointer + fixed stride` can perfectly match its semantic. I suggest providing both `base pointer + fixed stride` for ease-of-use and `array of pointers` for advanced and flexible usage.

Reply:
Echo @ZhennanQin .

Reply:
Having ptr+offsets help in the case of brgemm kernles when the work done by brgemm is small.
In depthwise conv (brdgmm) we cache the entire bactch (all combinations of bs_size, offsets required) and it seemed to have provided some perf benefit. See  https://github.com/intel-innersource/libraries.performance.math.onednn/pull/6220

But the argument of having constant pointers is counter-intutive to me... no benefit and limited flexibility.

Reply:
I don't thing there is any argument for having constant pointers, quite the opposite.
Offsets can be constants, but pointers will have to be recomputed always, even when offsets are constant.